Source/DafnyDriver/Commands/TranslateCommand.cs

```diff
@@ -16,7 +16,8 @@ static class TranslateCommand {
       CommonOptionBag.Output,
       IExecutableBackend.OuterModule,
       CommonOptionBag.IncludeRuntimeOption,
-      RunAllTestsMainMethod.IncludeTestRunner
+      RunAllTestsMainMethod.IncludeTestRunner,
+      CommonOptionBag.GenerateDoo
```

Comment:
If each translate invocation generates a separate doo file, then how do you get a doo file that contains information for different translates?

I thought the usage would be:

```
dafny build --target lib lib.dfy // produces lib.doo
dafny translate go lib.doo --update-doo // adds translation information for go to lib.doo
dafny translate cs lib.doo --update-doo // adds translation information for cs to lib.doo
```

Reply:
The idea is exactly that, to split files into target language doo files and not one master doo file. Tagging @robin-aws for more insight on that requirement.

Reply:
I think what we discussed what was that the `.doo` file could, in addition to what it contains now, also contain the contents of what an online repository for a Dafny library would have, which would be translation information for different targets and possibly references to the artifacts of those targets.

If we would be really persnickety, we could have a `.doo` file as it is now, with only information about the Dafny program, and a `.drep[ository]` file that contains the `.doo` and the translation information for different targets, but I don't see practical arguments for that amount of indirection.

Reply:
> ```
> dafny build --target lib lib.dfy // produces lib.doo
> dafny translate go lib.doo --update-doo // adds translation information for go to lib.doo
> dafny translate cs lib.doo --update-doo // adds translation information for cs to lib.doo
> ```

I think we do want to support this workflow somehow, eventually. But I see a couple of speed bumps that means this won't work that smoothly yet:

1. The Dafny pipeline doesn't really operate on a single input doo file. You can pass zero or more than one, so updating "the doo file" isn't the cleanest concept.
2. We don't yet have a good way of storing different values for different targets for options like `--outer-module` in a single doo file.

So I think it makes sense to have `--generate-doo` for now, and we can add `dafny doo merge ...` later to support the master doo file idea. I expect especially as we optimize the doo file format in the future, merging can become a fast operation.

Reply:
Alright, bear with me as I do a bit of a current state brain dump in order to help figure out the best way forwards here. I feel like I have only _part_ of a vision for how we support the build and packaging process for Dafny across all use cases and target languages, enough to feel able to offer counterpoints to some of the above, but not quite enough to offer alternatives I feel confident enough to propose yet. 

I think our focus should be to figure out the best first version of this feature that we're confident moves in a direction will support the eventual full solution, without having to wait to have consensus on what the full solution looks like, and without having to pull in more scope on this PR. And if we're asking for a version that has to change a lot from the current version, we should help make those changes.

I would in fact also like to clean up the implementation a little before merging as well, but that's a more minor point we can get to once we resolve this point of contention.

> I think what we discussed what was that the .doo file could, in addition to what it contains now, also contain the contents of what an online repository for a Dafny library would have, which would be translation information for different targets and possibly references to the artifacts of those targets.

I do remember that came up, but not that we agreed it was the only solution. My biggest worry about the master doo workflow is that it violates the pretty widespread idiom that build artifacts should be immutable assets, so that they can be aggressively cached by package managers and builds are easier to make reproducible. I don't have a concrete counterexample to show why we must not do this, but I also don't think we have enough alignment on what the actual Dafny build system/package manager will look like to know there isn't one.

> If we would be really persnickety, we could have a .doo file as it is now, with only information about the Dafny program, and a .drep[ository] file that contains the .doo and the translation information for different targets, but I don't see practical arguments for that amount of indirection.

The above might be an argument for that indirection, and something like this might turn out to be useful to integration with existing target language build systems better. My hope is that if adding translation information to .doo files is not flat out wrong or very expensive, it's an improvement over the current situation, and if we end up later on wanting to revert to the perspective that .doo files are only good for capturing the results of verification, that wouldn't be a big deal.

> To me that workflow seems more ambiguous, because the user now has multiple ways to generate a doo file, both through dafny build and dafny translate

This is fair, and when I suggested the `--generate-doo` approach I also suggested that I'd like to deprecate the "library backend". Implementing doo file building as a backend hasn't worked out very well in practice, since it really doesn't fit the shape of a backend in the implementation well. If nothing else, the fact that we ended up wanting to serialize the program before resolution, which we could only do by capturing the program earlier in the pipeline (the `ProgramAfterParsing` field) and referencing it from the LibraryBackend. 

So my full suggestion was moving doo file generation to somewhere more central, and instead of `dafny translate -t:lib` to build just a doo file with the results of verification, you could say `dafny verify --generate-doo`.

> Also, I feel generating the doo does not match with the spirit of dafny translate.

If we think of doo files as more generic Dafny Output Objects, and not just reusable library files, does it make more sense to say that any Dafny command should be able to produce them?

> Why would we not immediately add a target specific option section to doo files?

I think it's more complicated than that to be fully general, because we have options like `--unicode-char` and `--optimize-erasable-datatype-wrappers` that affect the translation output. That means if you intend to publish translated code for other consumers, you have to publish more than one version in a given target language, a separate version for each combination of such options they might want to use. 

Now it's natural to say we should drive towards a single support set of options and deprecate old ones, and I agree, but I also think we'll need to support the migratory period to get there well. It's also hard to be confident we're never going to want to introduce any similar options in the future, especially for important optimizations that otherwise break backwards compatibility of published translations.